ott/core/layers.py

```diff
@@ -0,0 +1,174 @@
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Lint as: python3
+"""Layers used in input convex neural networks (Amos+(2017), Bunne+(2022))."""
+
+from typing import Any, Callable, Tuple
+
+import jax
+import jax.numpy as jnp
+from flax import linen as nn
+
+PRNGKey = Any
+Shape = Tuple[int]
+Dtype = Any
+Array = Any
+
+
+class PositiveDense(nn.Module):
+  """A linear transformation using a weight matrix with all entries positive.
+
+  Args:
+    dim_hidden: the number of output dim_hidden.
+    beta: inverse temperature parameter of the softplus function (default: 1).
+    use_bias: whether to add a bias to the output (default: True).
+    dtype: the dtype of the computation (default: float32).
+    precision: numerical precision of computation see `jax.lax.Precision`
+      for details.
+    kernel_init: initializer function for the weight matrix.
+    bias_init: initializer function for the bias.
+  """
+  dim_hidden: int
+  beta: float = 1.0
+  use_bias: bool = True
+  dtype: Any = jnp.float32
+  precision: Any = None
+  kernel_init: Callable[
+      [PRNGKey, Shape, Dtype], Array] = nn.initializers.lecun_normal()
+  bias_init: Callable[[PRNGKey, Shape, Dtype], Array] = nn.initializers.zeros
+
+  @nn.compact
+  def __call__(self, inputs):
+    """Applies a linear transformation to inputs along the last dimension.
+
+    Args:
+      inputs: The nd-array to be transformed.
+    Returns:
+      The transformed input.
+    """
+    inputs = jnp.asarray(inputs, self.dtype)
+    kernel = self.param(
+        'kernel', self.kernel_init, (inputs.shape[-1], self.dim_hidden))
+    scaled_kernel = self.beta * kernel
+    kernel = jnp.asarray(
+        1 / self.beta * nn.softplus(scaled_kernel), self.dtype)
+    y = jax.lax.dot_general(
+        inputs, kernel, (((inputs.ndim - 1,), (0,)), ((), ())),
+        precision=self.precision)
+    if self.use_bias:
+      bias = self.param('bias', self.bias_init, (self.dim_hidden,))
+      bias = jnp.asarray(bias, self.dtype)
+      y = y + bias
+    return y
+
+
+class PosDefDense(nn.Module):
```

Comment:
I think PosDefDense is no longer needed.

Reply:
Removed.